src/reader.rs

```diff
@@ -88,31 +43,29 @@ where
     C: Read + Write,
 {
     async fn fill_buf(&mut self) -> Result<&[u8], ErrorKind> {
-        // We need to consume the loaded bytes before we read mode.
-        if self.buffer.is_empty() {
-            // embedded-tls has its own internal buffer, let's prefer that if we can
-            #[cfg(feature = "embedded-tls")]
-            if let HttpConnection::Tls(ref mut tls) = self.stream {
+        // The call to buffered.bypass() will only return Ok(...) if the buffer is empty.
+        // This ensures that we completely drain the possibly pre-filled buffer before we try
+        // to use the embedded-tls buffer directly.
+        // The matches/if let dance is to fix lifetime of the borrowed inner connection.
+        #[cfg(feature = "embedded-tls")]
+        if matches!(self.buffered.bypass(), Ok(HttpConnection::Tls(_))) {
+            if let HttpConnection::Tls(ref mut tls) = self.buffered.bypass().unwrap() {
```

Comment:
```suggestion
            if let Ok(HttpConnection::Tls(tls)) = self.buffered.bypass() {
```

This is just a bit of stylistic thing, though, doesn't solve our lifetime problem. I expect we'd have to wait for Polonius for that to change.